File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/payment/application/command/CreatePaymentService.java
package com.example.accountservice.payment.application.command;

import com.example.accountservice.payment.application.adapter.AccountAdapter;
import com.example.accountservice.payment.application.adapter.CardAdapter;
import com.example.accountservice.payment.application.query.GetPaymentResult;
import com.example.accountservice.payment.domain.Payment;
import com.example.accountservice.payment.domain.PaymentException;
import com.example.accountservice.payment.domain.PaymentRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class CreatePaymentService {

    private final PaymentRepository paymentRepository;
    private final CardAdapter cardAdapter;
    private final AccountAdapter accountAdapter;

    public GetPaymentResult create(CreatePaymentCommand command) {
        // Confirm via the synchronous Adapter (ACL) that the card exists and is active — a
        // synchronous call is needed because the response (whether the payment can proceed)
        // depends on it.
        CardAdapter.CardView card =
                cardAdapter
                        .findCard(command.cardId(), command.requesterId())
                        .orElseThrow(
                                () ->
                                        new PaymentException(
                                                PaymentException.ErrorCode.LINKED_CARD_NOT_FOUND,
                                                "The card to link could not be found."));
        if (!card.active()) {
            throw new PaymentException(
                    PaymentException.ErrorCode.PAYMENT_REQUIRES_ACTIVE_CARD,
                    "Only an active card can be used for payment.");
        }

        // Confirm via the synchronous Adapter (ACL) that the linked account is active and has a
        // sufficient balance (a read-only judgment). The actual deduction does not happen here —
        // the Account BC subscribes to PaymentCompletedEvent -> payment.completed.v1 and performs
        // it asynchronously (the "sync=query, async=state change" principle in cross-domain.md).
        AccountAdapter.AccountView account =
                accountAdapter
                        .findAccount(card.accountId(), command.requesterId())
                        .orElseThrow(
                                () ->
                                        new PaymentException(
                                                PaymentException.ErrorCode.LINKED_ACCOUNT_NOT_FOUND,
                                                "The linked account could not be found."));
        if (!account.active()) {
            throw new PaymentException(
                    PaymentException.ErrorCode.PAYMENT_REQUIRES_ACTIVE_ACCOUNT,
                    "Only an active account can be used for payment.");
        }
        if (account.balanceAmount() < command.amount()) {
            throw new PaymentException(
                    PaymentException.ErrorCode.INSUFFICIENT_BALANCE,
                    "Cannot make the payment because the account balance is insufficient.");
        }

        Payment payment =
                Payment.create(
                        command.cardId(),
                        card.accountId(),
                        command.requesterId(),
                        command.amount());
        payment.complete();

        paymentRepository.savePayment(payment);
        return GetPaymentResult.from(payment);
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/payment/application/command/CreatePaymentService.kt
package com.example.accountservice.payment.application.command

import com.example.accountservice.payment.application.adapter.AccountAdapter
import com.example.accountservice.payment.application.adapter.CardAdapter
import com.example.accountservice.payment.domain.InsufficientBalanceException
import com.example.accountservice.payment.domain.LinkedAccountNotFoundException
import com.example.accountservice.payment.domain.LinkedCardNotFoundException
import com.example.accountservice.payment.domain.Payment
import com.example.accountservice.payment.domain.PaymentRepository
import com.example.accountservice.payment.domain.PaymentRequiresActiveAccountException
import com.example.accountservice.payment.domain.PaymentRequiresActiveCardException
import org.springframework.stereotype.Service

/**
 * Processes a payment (card + balance verification) — a combination of Adapters (synchronous reads),
 * not a Domain Service.
 *
 * 1. Synchronous Adapter → Card BC: verifies the card is ACTIVE and confirms the linked accountId
 *    (reusing the same ACL pattern that Card already uses to query Account).
 * 2. Synchronous Adapter → Account BC: verifies balance ≥ payment amount (a read-only judgment,
 *    "whether payment is possible").
 * 3. If both pass, creates the Payment and immediately transitions it to COMPLETED, publishing
 *    PaymentCompletedEvent (a Domain Event) and loading it into the Outbox.
 *
 * The actual balance deduction is not performed by this synchronous call — Account BC subscribes to
 * the `payment.completed.v1` Integration Event and performs it asynchronously (the "synchronous =
 * query, asynchronous Integration Event = state change" principle from cross-domain.md).
 */
@Service
class CreatePaymentService(
    private val paymentRepository: PaymentRepository,
    private val cardAdapter: CardAdapter,
    private val accountAdapter: AccountAdapter,
) {
    fun create(command: CreatePaymentCommand): CreatePaymentResult {
        val card = cardAdapter.findCard(command.cardId, command.requesterId) ?: throw LinkedCardNotFoundException()
        if (!card.active) throw PaymentRequiresActiveCardException()

        val account =
            accountAdapter.findAccount(card.accountId, command.requesterId) ?: throw LinkedAccountNotFoundException()
        if (!account.active) throw PaymentRequiresActiveAccountException()
        if (account.balanceAmount < command.amount) throw InsufficientBalanceException()

        val payment =
            Payment.create(cardId = command.cardId, accountId = card.accountId, ownerId = command.requesterId, amount = command.amount)
        payment.complete()

        paymentRepository.savePayment(payment)

        return CreatePaymentResult(
            paymentId = payment.paymentId,
            cardId = payment.cardId,
            accountId = payment.accountId,
            ownerId = payment.ownerId,
            amount = payment.amount,
            status = payment.status.name,
            createdAt = payment.createdAt,
        )
    }
}
